### PR TITLE
feat(readableColor): add strict mode option to meet WCAG AA

### LINKED ISSUE #479: Enhance `readableColor` to support preferences
We've been using Polished more and more in our current project, and find it particularly useful for accessibility.

**The problem**

`readableColor` will return a color that is not accessible whenever the user-supplied `lightReturnColor`/`darkReturnColor` does not contrast strongly enough with `color`.

This is not an issue if the user does not supply `lightReturnColor`/`darkReturnColor` as black/white are returned.

**Proposal**

To resolve this, I propose that we add `strict` flag as a fourth parameter, that ensures the function always returns an accessible color (as per WCAG AA).

```ts
readableColor(color: string, lightReturnColor: string, darkReturnColor: string, strict: boolean)
```

With this flag, if either`lightReturnColor` or `darkReturnColor` do not meet contrast requirements against a given color, `black` or `white` would be returned instead.

This would not be a breaking change.

**Possible change for next major**

In the next major release of polished, this _could_ become the default setting and `strict` could be removed.

---

If you think this change is acceptable, I'll raise a PR. Feedback and other thoughts welcome.

### PR DESCRIPTION
**Pull request by mrmckeb**
Resolves #479

### CODE REVIEW

File: src/color/readableColor.js [hunk 1]
```diff
@@ -1,8 +1,15 @@
...
+import meetsContrastGuidelines from './meetsContrastGuidelines'
 
 /**
- * Returns black or white (or optional light and dark return colors) for best contrast depending on the luminosity of the given color.
+ * Returns black or white (or optional light and dark return colors) for best
+ * contrast depending on the luminosity of the given color.
+ *
+ * Set `strict` to `true` to ensure that the return color always meets or
+ * exceeds WCAG level AA or greater. In this case, black or white are used as
+ * fallback colors.
```
Comment (bhough) [inline src/color/readableColor.js]: Let's update this to make it clear that this only applies when passing custom colors.

File: src/color/readableColor.js [hunk 2]
```diff
@@ -26,13 +35,30 @@ import getLuminance from './getLuminance'
...
+const defaultLightReturnColor = '#000'
+const defaultDarkReturnColor = '#fff'
+
 export default function readableColor(
   color: string,
-  lightReturnColor: string = '#000',
-  darkReturnColor: string = '#fff',
+  lightReturnColor: string = defaultLightReturnColor,
+  darkReturnColor: string = defaultDarkReturnColor,
+  strict: boolean = false,
```
Comment (bhough) [inline src/color/readableColor.js]: Not something you introduced, but let's notate these as optional `?` for Flow.

File: src/color/readableColor.js [hunk 3]
```diff
@@ -26,13 +35,30 @@ import getLuminance from './getLuminance'
...
+
+  // May return a color that does not meet WCAG AA.
+  // TODO: Make `strict` the default behaviour in the next major release.
+  if (!strict) {
+    return preferredReturnColor
+  }
+
+  if (meetsContrastGuidelines(color, preferredReturnColor).AA) {
+    return preferredReturnColor
+  }
```
Comment (bhough) [inline src/color/readableColor.js]: You should be able to combine these two checks into an `||` check to reduce verbosity.

File: src/color/readableColor.js [hunk 4]
```diff
@@ -26,13 +35,30 @@ import getLuminance from './getLuminance'
...
+  const isLightColor = getLuminance(color) > 0.179
+  const preferredReturnColor = isLightColor ? lightReturnColor : darkReturnColor
+
+  // May return a color that does not meet WCAG AA.
+  // TODO: Make `strict` the default behaviour in the next major release.
+  if (!strict) {
+    return preferredReturnColor
+  }
+
+  if (meetsContrastGuidelines(color, preferredReturnColor).AA) {
```
Comment (bhough) [inline src/color/readableColor.js]: Pulling in `meetsContrastGuidelines` will include a chain of modules that will increase the size of `readableColor`. We might be able to go down the chain a bit (for example just check for the AA value against `getContrast`) to reduce the size some.

File: src/color/readableColor.js (hunk 2 above)
Comment (mrmckeb) [reply src/color/readableColor.js]: I work with TypeScript, and thought this looked wrong - but lazily assumed there was something flow-related going on. Sorry, updated!

File: src/color/readableColor.js (hunk 3 above)
Comment (mrmckeb) [reply src/color/readableColor.js]: I'll take a look - I was trying to balance verbosity and readability. I'm sure I can clean it up more.

File: src/color/readableColor.js (hunk 4 above)
Comment (mrmckeb) [reply src/color/readableColor.js]: Agreed. I also thought about this, but was also concerned with having the same logic living in multiple places.

However, the AA standard is unlikely to change.

**Comment by mrmckeb** [comment]
Thanks again!

**Pull request merged**